deployer/scripts/upgrade.sh

```diff
@@ -278,20 +286,102 @@ function getArrayIndex() {
   fi
 }
 
+# This function will wait until the latestVersion changes
+function waitForChange() {
+
+  local currentVersion=$1
+  local dc=$2
+
+  waitFor "[[ $currentVersion -lt \$(oc get $dc -o jsonpath='{.status.latestVersion}') ]]" && return 0
+
+  return 1
+}
+
+# This function will tell us if the proposed patches are different than the current values
+function isValidChange() {
+
+  local object=$1; shift
+
+  for patch in "$@"; do
+    path="$(echo $patch | cut -d"=" -f 1)"
+    value="$(echo $patch | cut -d"=" -f 2)"
+    result="$(oc get $object -o jsonpath="$path")"
+
+    # we want to do a conversion of the image tag to the sha256 when checking images
+    # if path contains "containers" then we're doing an image patch
+    # the IS is between a "/" and ":" in value
+    if [[ "$path" =~ "containers" ]]; then
+      #we want to check if the tag or the sha256 match
+      image=$(echo $value | sed 's/^.*\///')
+
+      is=$(echo $image | cut -d":" -f 1)
+      tag=$(echo $image | cut -d":" -f 2)
+
+      shaValue=$(echo $value | grep -o '^.*:')
+      # drop last ":" and replace with "@"
+      shaValue=$(echo ${shaValue:0:-1}@)
+      shaValue+=$(oc get is $is -o jsonpath="{.status.tags[?(@.tag==\"$tag\")].items[0].image}")
+
+      [[ "$result" != "$value" ]] && \
+      [[ "$result" != "$shaValue" ]] && return 0
+    else
+      [[ "$result" != "$value" ]] && return 0
+    fi
+  done
+
+  echo "Same values as intended patches for $object"
+  return 1
+}
+
+function patchIfValid() {
+
+  local object=$1; shift
+  local isDC=false
+  local currentVersion
+  local actualPatch=()
+
+  [[ -n $(echo $object | grep '^dc\/') ]] && isDC=true
+
+  if [ $isDC = true ]; then
+    currentVersion=$(oc get $object -o jsonpath='{.status.latestVersion}')
+  fi
+
+  if ! isValidChange "$object" "$@"; then
+    return 0
+  fi
+
+  for patch in "$@"; do
+    # delimeter is '='
+    path=$(echo $patch | cut -d"=" -f 1 | sed 's/[\.\[]/\//g ; s/\]//g ; s/[\}\{]//g')
+    value=$(echo $patch | cut -d"=" -f 2)
+
+    actualPatch+=( "{\"op\":\"replace\",\"path\":\"$path\",\"value\":\"$value\"}")
```

Comment:
That works fine as long as there are no spaces in $path or $value, which as
you can tell i hate to rely on...

On Tue, May 24, 2016 at 3:25 PM, Eric Wolinetz <notifications@github.com>
wrote:

> In deployer/scripts/upgrade.sh
> <https://github.com/openshift/origin-aggregated-logging/pull/141#discussion_r64454371>
> :
>
> > +  [[ -n $(echo $object | grep '^dc\/') ]] && isDC=true
> > +
> > +  if [ $isDC = true ]; then
> > +    currentVersion=$(oc get $object -o jsonpath='{.status.latestVersion}')
> > +  fi
> > +
> > +  if ! isValidChange "$object" "$@"; then
> > +    return 0
> > +  fi
> > +
> > +  for patch in "$@"; do
> > +    # delimeter is '='
> > +    path=$(echo $patch | cut -d"=" -f 1 | sed 's/[\.\[]/\//g ; s/\]//g ; s/[\}\{]//g')
> > +    value=$(echo $patch | cut -d"=" -f 2)
> > +
> > +    actualPatch+=( "{\"op\":\"replace\",\"path\":\"$path\",\"value\":\"$value\"}")
>
> is there a reason why we'd put quotes around $path and $value? couldn't
> we just have
>
> actualPatch+=( '{"op":"replace","path":"'$path'","value":"'$value'"}' )
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin-aggregated-logging/pull/141/files/bba90e8f1016b3bd4df104fc0c35fddfe93773b7#r64454371>
>